gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -0,0 +1,274 @@
+import gobblin.configuration.SourceState;
+import gobblin.configuration.WorkUnitState;
+import gobblin.source.extractor.Extractor;
+import gobblin.source.extractor.extract.MessageBasedSource;
+import gobblin.source.workunit.Extract;
+import gobblin.source.workunit.WorkUnit;
+import gobblin.source.workunit.Extract.TableType;
+
+import com.google.common.base.Splitter;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.io.Closer;
+
+
+/**
+ * Kafka source implementation
+ *
+ * @author ziliu
+ *
+ */
+public class KafkaSource extends MessageBasedSource<Schema, GenericRecord> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaSource.class);
+
+  public static final String TOPIC_BLACKLIST = "topic.blacklist";
+  public static final String TOPIC_WHITELIST = "topic.whitelist";
+  public static final String TOPICS_MOVE_TO_LATEST_OFFSET = "topics.move.to.latest.offset";
+  public static final String MOVE_TO_EARLIEST_OFFSET_ALLOWED = "topics.move.to.earliest.offset";
+  public static final boolean DEFAULT_MOVE_TO_EARLIEST_OFFSET_ALLOWED = false;
+
+  public static final String TOPIC_NAME = "topic.name";
+  public static final String PARTITION_ID = "partition.id";
+
+  private final Set<String> moveToLatestTopics = Sets.newTreeSet(String.CASE_INSENSITIVE_ORDER);
+  private final Map<KafkaPartition, Long> previousOffsets = Maps.newHashMap();
+
+  @Override
+  public List<WorkUnit> getWorkunits(SourceState state) {
+    List<WorkUnit> workUnits = Lists.newArrayList();
+    Closer closer = Closer.create();
+    try {
+      KafkaWrapper kafkaWrapper = closer.register(KafkaWrapper.create(state));
+      List<KafkaTopic> topics = getFilteredTopics(kafkaWrapper, state);
+      for (KafkaTopic topic : topics) {
+        workUnits.addAll(getWorkUnitsForTopic(kafkaWrapper, topic, state));
+      }
+      return workUnits;
+    } finally {
+      try {
+        closer.close();
+      } catch (IOException e) {
+        LOG.error("Failed to close kafkaWrapper", e);
+      }
+    }
+  }
+
+  private List<WorkUnit> getWorkUnitsForTopic(KafkaWrapper kafkaWrapper, KafkaTopic topic, SourceState state) {
+    List<WorkUnit> workUnits = Lists.newArrayList();
+    for (KafkaPartition partition : topic.getPartitions()) {
+      WorkUnit workUnit = getWorkUnitForTopicPartition(kafkaWrapper, partition, state);
+      if (workUnit != null) {
+        workUnits.add(workUnit);
+      }
+    }
+    return workUnits;
+  }
+
+  private WorkUnit getWorkUnitForTopicPartition(KafkaWrapper kafkaWrapper, KafkaPartition partition, SourceState state) {
+    Offsets offsets = new Offsets();
+
+    try {
+      offsets.setEarliestOffset(kafkaWrapper.getEarliestOffset(partition));
+      offsets.setLatestOffset(kafkaWrapper.getLatestOffset(partition));
+      if (shouldMoveToLatestOffset(partition, state)) {
+        offsets.startAtLatestOffset();
```

Comment:
Yes, it is used for bootstrapping. Topics that move to the latest offset will not be consumed in this run, but in the next run, it is supposed to be removed from the list and will then be consumed.